File: mqtt/golang/controllers/mqtt_controller.go
package controllers

import (
	"golang-note-mqtt/services"
	"net/http"

	"github.com/labstack/echo/v4"
)

type MqttController interface {
	SendMessage(c echo.Context) error
}

type mqttController struct {
	MqttService services.MqttService
}

func NewMqttController(mqttService services.MqttService) MqttController {
	return &mqttController{
		MqttService: mqttService,
	}
}

func (controller *mqttController) SendMessage(c echo.Context) error {
	message := c.QueryParam("message")
	response := controller.MqttService.SendMessage(message)
	return c.JSON(http.StatusOK, map[string]interface{}{
		"response": response,
	})
}

File: mqtt/rust/src/controllers/mqtt_controller.rs
use std::sync::Arc;

use axum::{extract::{Path, State}, http::StatusCode, response::IntoResponse};

use crate::{services::mqtt_service::MqttService, states::app_state::AppState};

pub async fn send_message(
    State(state): State<Arc<AppState>>,
    Path(message): Path<String>
) -> impl IntoResponse {
    let response = state.mqtt_service.send_message(message).await;
    (StatusCode::OK, response)
}